### ISSUE TITLE
How to artificially increase dictionary size for fine-tuning Tacotron2V4

### ISSUE THREAD

**Issue by ZDisket** [espnet/espnet#1975]
I am trying to train a phonetic model by fine-tuning the 0.6.1 release LJSpeech Tacotron2v4 model. After I had added the code for transfer learning found [here](https://github.com/espnet/espnet/issues/1208), I came across a `size mismatch for enc.embed.weight` error. Figured out it was the dictionary size so I removed some stuff from the transcripts and it went well. 
I tried the same again but this time the dictionary was too small, so first I tried adding some extra stuff to one of the transcriptions (data/phn_train/text) to match the size. It was all removed in stage 2 of run.sh. I have no idea how it can tell it's been artificially added.
![image](https://user-images.githubusercontent.com/30500847/82984528-bd835d80-9fc8-11ea-9ed5-4a10a965b2ce.png)
Next thing was going to data/lang_1phn/phn_train_no_dev_units.txt and adding the additional phenomenes and incremental IDs myself. When I started training, it still gave me `size mismatch for enc.embed.weight: copying a param with shape torch.Size([77, 512]) from checkpoint, the shape in current model is torch.Size([73, 512]).` 

So my question is, **how do I artificially increase the dictionary size so the pretrained model loads well?**

**Comment by unilight** [comment]
I think there is no simple hack to your problem. First you will have to be careful about the ordering of your original and augmented dictionary, so that the newly added phonemes do not affect the original IDs. Then, you will need to modify the code, so you can load the pretrained `[73, 512]` embedding table into the proper position of the new `[77, 512]` table.
My suggestion is that if the phonemes that only appear in your adaptation data do not take a big proportion, then just leave them as OOV and train to see what happens. The newly added phonemes are learned using the adaptation data only and might not guarantee to generalize well. If the result is bad, then consider the above troublesome approach.

**Comment by kan-bayashi** [comment]
Another easier choice is fine-tuning with the initialized embed layer.
In this case, you can load the parameters of the pretrained model except for the embedding layer of the encoder and then perform fine-tuning.
See the following documents to load the subset of the parameters of the pretrained model.
https://github.com/espnet/espnet/blob/develop/doc/tutorial.md#how-to-use-finetuning

**Comment by ZDisket** [comment]
> Another easier choice is fine-tuning with the initialized embed layer.
> In this case, you can load the parameters of the pretrained model except for the embedding layer of the encoder and then perform fine-tuning.
> See the following documents to load the subset of the parameters of the pretrained model.
> https://github.com/espnet/espnet/blob/develop/doc/tutorial.md#how-to-use-finetuning

![image](https://user-images.githubusercontent.com/30500847/83038237-1c6dc480-a013-11ea-9ab3-0b098660f183.png)

**Comment by b-flo** [comment]
Hi,

It's an option for {asr, tts}_train.py, or it can be put in your training config ("enc-init: model/snapshot"). Maybe I should stipulate that in the doc.

**Comment by ZDisket** [comment]
@b-flo 
> It's an option for {asr, tts}_train.py, or it can be put in your training config ("enc-init: model/snapshot").
I cloned the 0.6.3 release for the options and added some to the YAML, now it's training well. 
Thank you!

**Comment by ZDisket** [comment]
Now the model fails to learn attention and spouts [pure nonsense.](https://voca.ro/2HEFvVDJUBX) Any other ideas? Why won't run.sh recognize my new data/lang_1phn/phn_train_no_dev_units.txt? My `enc_init_mods` are `enc.0.,enc.1.,enc.2.`, maybe I missed something?
![badalign](https://user-images.githubusercontent.com/30500847/83085787-4c44b880-a063-11ea-892e-275f9e6c2352.gif)

**Comment by kan-bayashi** [comment]
Did you also use `dec-init`?

**Comment by ZDisket** [comment]
> Did you also use `dec-init`?
Yes. I have both `dec-init` and `enc-init` in the YAML set to the pretrained model. I've got the 0.6.3 release and applying on the 0.6.0 model found in the Google Drive if that helps.

**Comment by kan-bayashi** [comment]
Could you attach the whole your config?

**Comment by ZDisket** [comment]
> Could you attach the whole your config?

I don't have it since I run it on Colaboratory, but this is the code that generates it. 
![image](https://user-images.githubusercontent.com/30500847/83086492-44861380-a065-11ea-8cae-cad76455d554.png)
But your comment made me inspect the train.log in case my code was broken somehow and found this 

> WARNING: reading model parameters from pretr.avg.best
> 2020-05-27 16:03:53,483 (asr_init:114) WARNING: module(s) ['enc.0.', 'enc.1.', 'enc.2.'] don't match or (partially match) available modules in model.
> 2020-05-27 16:03:53,484 (asr_init:115) WARNING: for information, the existing modules in model are:
> 2020-05-27 16:03:53,484 (asr_init:116) WARNING: ['enc.embed.weight', 'enc.convs.0.0.weight', 'enc.convs.0.1.weight', 'enc.convs.0.1.bias', 'enc.convs.0.1.running_mean', 'enc.convs.0.1.running_var', 'enc.convs.0.1.num_batches_tracked', 'enc.convs.1.0.weight', 'enc.convs.1.1.weight', 'enc.convs.1.1.bias', 'enc.convs.1.1.running_mean', 'enc.convs.1.1.running_var', 'enc.convs.1.1.num_batches_tracked', 'enc.convs.2.0.weight', 'enc.convs.2.1.weight', 'enc.convs.2.1.bias', 'enc.convs.2.1.running_mean', 'enc.convs.2.1.running_var', 'enc.convs.2.1.num_batches_tracked', 'enc.blstm.weight_ih_l0', 'enc.blstm.weight_hh_l0', 'enc.blstm.bias_ih_l0', 'enc.blstm.bias_hh_l0', 'enc.blstm.weight_ih_l0_reverse', 'enc.blstm.weight_hh_l0_reverse', 'enc.blstm.bias_ih_l0_reverse', 'enc.blstm.bias_hh_l0_reverse', 'dec.att.mlp_enc.weight', 'dec.att.mlp_enc.bias', 'dec.att.mlp_dec.weight', 'dec.att.mlp_att.weight', 'dec.att.loc_conv.weight', 'dec.att.gvec.weight', 'dec.att.gvec.bias', 'dec.lstm.0.cell.weight_ih', 'dec.lstm.0.cell.weight_hh', 'dec.lstm.0.cell.bias_ih', 'dec.lstm.0.cell.bias_hh', 'dec.lstm.1.cell.weight_ih', 'dec.lstm.1.cell.weight_hh', 'dec.lstm.1.cell.bias_ih', 'dec.lstm.1.cell.bias_hh', 'dec.prenet.prenet.0.0.weight', 'dec.prenet.prenet.0.0.bias', 'dec.prenet.prenet.1.0.weight', 'dec.prenet.prenet.1.0.bias', 'dec.postnet.postnet.0.0.weight', 'dec.postnet.postnet.0.1.weight', 'dec.postnet.postnet.0.1.bias', 'dec.postnet.postnet.0.1.running_mean', 'dec.postnet.postnet.0.1.running_var', 'dec.postnet.postnet.0.1.num_batches_tracked', 'dec.postnet.postnet.1.0.weight', 'dec.postnet.postnet.1.1.weight', 'dec.postnet.postnet.1.1.bias', 'dec.postnet.postnet.1.1.running_mean', 'dec.postnet.postnet.1.1.running_var', 'dec.postnet.postnet.1.1.num_batches_tracked', 'dec.postnet.postnet.2.0.weight', 'dec.postnet.postnet.2.1.weight', 'dec.postnet.postnet.2.1.bias', 'dec.postnet.postnet.2.1.running_mean', 'dec.postnet.postnet.2.1.running_var', 'dec.postnet.postnet.2.1.num_batches_tracked', 'dec.postnet.postnet.3.0.weight', 'dec.postnet.postnet.3.1.weight', 'dec.postnet.postnet.3.1.bias', 'dec.postnet.postnet.3.1.running_mean', 'dec.postnet.postnet.3.1.running_var', 'dec.postnet.postnet.3.1.num_batches_tracked', 'dec.postnet.postnet.4.0.weight', 'dec.postnet.postnet.4.1.weight', 'dec.postnet.postnet.4.1.bias', 'dec.postnet.postnet.4.1.running_mean', 'dec.postnet.postnet.4.1.running_var', 'dec.postnet.postnet.4.1.num_batches_tracked', 'dec.feat_out.weight', 'dec.prob_out.weight', 'dec.prob_out.bias', 'taco2_loss.bce_criterion.pos_weight']

So I'll try again and see what happens.

**Comment by kan-bayashi** [comment]
Oh, it seems that none of the parameters are loaded.
You should specify like `enc.convs.,enc.blstm.,`.

**Comment by ZDisket** [comment]
Alrighty, it's now training well. Thanks!

**Comment by ShreyaKhare** [comment]
@ZDisket @kan-bayashi  Could please elaborate how did this work ? 
I am trying to do the same for transformer model. 
I am getting this warning  WARNING: modules ['decoder.'] in model snapshot.ep.20 don't match your training config ..

**Comment by kan-bayashi** [comment]
@ShreyaKhare Please write the details of your procedure.
(e.g. your config, all of the log, and snippet to reproduce your problem)

**Comment by Khelp** [comment]
Dear @kan-bayashi ,
I have the sample problem as @ShreyaKhare .
For configuration, I follow the default configuration of the ESPnet Transformer.
In the ASR training script, I have added
```
--enc-init exp/*/model.acc.best
--dec-init exp/*/model.acc.best
--enc-init-mods encoder.
--dec-init-mods decoder.
--freeze-mods encoder.embed.
```
Here is the log:
```
WARNING: loading ['encoder.'] from model: exp/*/model.acc.best
(asr_init:239) WARNING: override encoder.embed.conv.0.weight
(asr_init:239) WARNING: override encoder.embed.conv.0.bias
(asr_init:239) WARNING: override encoder.embed.conv.2.weight
(asr_init:239) WARNING: override encoder.embed.conv.2.bias
(asr_init:239) WARNING: override encoder.embed.out.0.weight
(asr_init:239) WARNING: override encoder.embed.out.0.bias
(asr_init:239) WARNING: override encoder.encoders.0.self_attn.linear_q.weight
......
WARNING: reading model parameters from exp/*/model.acc.best
(ctc:36) WARNING: CTC was set to builtin due to PyTorch version.
(asr_init:280) WARNING: modules ['decoder.'] in model exp/*/model.acc.best don't match your training config
```
Is there any problem in configuration or do I need to modify some part of the program?
Best,

**Comment by kan-bayashi** [comment]
Please make sure the decoder architecture is the same as the pretrained model.

**Comment by Khelp** [comment]
Yes, actually, I don't change anything in the decoding part.

**Comment by kan-bayashi** [comment]
In that case, the size of the output layer is different. That causes the mismatch.

**Comment by Khelp** [comment]
Dear @kan-bayashi,
My target is to improve ASR performance by using the pre-trained model.
I also want to extend the vocabulary size.
So, could you give me a suggestion?

**Comment by ShreyaKhare** [comment]
@Khelp  I am also facing the same issue , I was able to resolve it using this config : 
enc-init: exp/*/model.loss.best
enc-init-mods: 'encoder.'
dec-init: exp/*/model.loss.best
dec-init-mods: 'decoder.decoders.'

However, I also want to try increasing my dictionary size while finetuning, what is the best way to do that ? 
@kan-bayashi

**Comment by kan-bayashi** [comment]
Your config seems to be OK but maybe `decoder.after_norm` is missing.
```
enc-init-mods: encoder.embed
dec-init-mods: decoder.decoders,decoder.after_norm
```

**Comment by ShreyaKhare** [comment]
Thanks @kan-bayashi for pointing this out .

**Comment by Khelp** [comment]
Dear @kan-bayashi,
I have one more problem with the "main loss" value.
The main loss value of the pre-trained model is less than 10, but after tunning around 30 epochs, I found that the loss value is very high. It's bigger than 40.
So, what is the problem with this?

**Comment by kan-bayashi** [comment]
@Khelp The question is too general since there are a lot of possible reasons (the adaptation dataset size, vocabulary, etc...). So I have no clear answer to your question.
If you want to discuss more, please open the issue with detailed settings.
E.g.:
- Pretrained model detail
- Adaptation configuration
- Dataset detail
- Training curve
